src/main/scala/mesosphere/marathon/state/Migration.scala

```diff
@@ -367,6 +374,129 @@ class MigrationTo0_16(groupRepository: GroupRepository, appRepository: AppReposi
   }
 }
 
+class MigrationTo1_1(groupRepository: GroupRepository, appRepository: AppRepository, conf: MarathonConf) {
+  private[this] val log = LoggerFactory.getLogger(getClass)
+
+  def migrate(): Future[Unit] = {
+    log.info("Start 1.1 migration")
+
+    //        We can have 3 cases here:
+    //          1. App has one entry at the wrong place and must be moved:
+    //
+    //              Group( id = /, apps = [“/foo/bla”] )
+    //              —>
+    //              Group ( id = / ,
+    //                Group( id = /foo, apps = [”/foo/bla”] )
+    //
+    //          2. App has 2 entries but both entries has the same version so the wrong one can be deleted:
+    //
+    //            Group( id = /, apps = [“/foo/bla, version = 1"],
+    //              Group( id = /foo, apps = [“/foo/bla, version = 1"])
+    //            —>
+    //            Group ( id = / ,
+    //              Group( id = /foo, apps = [“/foo/bla, version = 1"] )
+    //
+    //          3. App has 2 entries with different versions. At this point we can either fail the migration or keep one of
+    //             the versions. After some discussion we decided to keep the latest version.
+    //             CAUTION: this migration can potentially result in data loss:
+    //
+    //            Group( id = /, apps = [“/foo/bla, version = 1"],
+    //              Group( id = /foo, apps = [“/foo/bla, version = 2"])
+    //            —>
+    //            Group ( id = / ,
+    //              Group( id = /foo, apps = [“/foo/bla, version = 2"] )
+
+    val id = groupRepository.zkRootName
+
+    for {
+      updatedGroups <- updateGroups(id)             // Update root and it's versions
+      _ = log.info(s"Updated groups: $updatedGroups")
+      _ <- storeGroups(id, updatedGroups)           // Store updated groups
+      _ <- updateApps()                             // Update apps from the root group
+    } yield log.info("Finished 1.1 migration")
+  }
+
+  def updateGroups(id: String): Future[Iterable[Group]] = {
+    groupRepository.listVersions(id).flatMap { versions =>
+      val fs = versions.map(version =>
+        groupRepository.group(id, version).map {
+          case Some(group) =>
+            log.debug(s"Loaded group: $group")
+            val updated = validateGroup(updateGroup(group))
+            log.debug(s"Updated group: $updated")
+            updated
+          case None => throw new MigrationFailedException(s"Group $id:$version not found")
+        }
+      )
+      Future.sequence(fs)
+    }
+  }
+
+  def storeGroups(id: String, updatedGroups: Iterable[Group]): Future[Iterable[Group]] = {
+    val storedGroups = updatedGroups.map { group =>
+      groupRepository.store(id, group)
```

Comment:
for some reason I keep mix them up. yes, I meant the newest. Versions are sorted from the **oldest** to the **newest** so that the newest is stored last. I read this sentence twice to be sure ^^